src/sage/numerical/backends/matrix_backend.pyx

```diff
@@ -0,0 +1,1170 @@
+            sage: p.ncols()
+            0
+            sage: p.nrows()
+            0
+            sage: p.add_linear_constraints(5, 0, None)
+            sage: p.add_col(range(5), range(5))
+            sage: p.nrows()
+            5
+        """
+        column = []
+        for _ in indices:
+            column.append(self._base_ring.zero())
+
+        for idx, ind in enumerate(indices):
+            column[ind] = coeffs[idx]
+
+        if self.ncols() == 0:
+            self.Matrix = matrix(len(column), column)
+        else:
+            self.Matrix = self.Matrix.augment(matrix(len(column), column))
+
+        self.col_lower_bound_indicator.append(None)
+        if self.col_lower_bound.dimensions()[1] == 0:
+            self.col_lower_bound = matrix(1, [self._base_ring.zero()])
+        else:
+            self.col_lower_bound = self.col_lower_bound.augment(matrix([self._base_ring.zero()]))
+
+        self.col_upper_bound_indicator.append(None)
+        if self.col_upper_bound.dimensions()[1] == 0:
+            self.col_upper_bound = matrix(1, [self._base_ring.zero()])
+        else:
+            self.col_upper_bound = self.col_upper_bound.augment(matrix([self._base_ring.zero()]))
+            
+        self.objective_coefficients = self.objective_coefficients.augment(matrix([self._base_ring.zero()]))
+        self.col_name_var.append(None)
+
+    cpdef add_linear_constraint(self, coefficients, lower_bound, upper_bound, name=None):
+        """
+        Add a linear constraint.
+
+        INPUT:
+
+        - ``coefficients`` an iterable with ``(c,v)`` pairs where ``c``
+          is a variable index (integer) and ``v`` is a value (real
+          value).
+
+        - ``lower_bound`` - a lower bound, either a real value or ``None``
+
+        - ``upper_bound`` - an upper bound, either a real value or ``None``
+
+        - ``name`` - an optional name for this row (default: ``None``)
+
+        EXAMPLES::
+
+            sage: from sage.numerical.backends.generic_backend import get_solver
+            sage: p = get_solver(solver = "Matrix")                 
+            sage: p.add_variables(5)                                
+            4
+            sage: p.add_linear_constraint(zip(range(5), range(5)), 2.0, 2.0)                
+            sage: p.row(0)                                          
+            ([1, 2, 3, 4], [1, 2, 3, 4])
+            sage: p.row_bounds(0)                                   
+            (2, 2)
+            sage: p.add_linear_constraint(zip(range(5), range(5)), 1.0, 1.0, name='foo')    
+            sage: p.row_name(-1)                                    
+            'foo'
+        """
+        coefficients = list(coefficients)
+        for c in coefficients:
+            while c[0] > self.ncols() - 1:
+                self.add_variable()
+                     
+        if self.Matrix.dimensions()[0] == 0:
+            self.Matrix = matrix(1, [self._base_ring.zero() for i in range(len(coefficients))])
```

Comment:
`len(coefficients)` is wrong; you need a vector of length self.Matrix.dimensions()[1].

Add a doctest with an actual sparse vector as `coeffients`, for example [(1, 7), (3, 10)]